### ISSUE TITLE
ocamlfind package assumes ocamlopt presence

### ISSUE THREAD

**Issue by smondet** [ocaml/opam-repository#1680]
The package calls `make opt` without checking whether it is a bytecode-only platform.

https://github.com/ocaml/opam-repository/blob/master/packages/ocamlfind/ocamlfind.1.4.0/opam

PS: is there an Opam variable or something to fix that?
PPS: this was found on `#ocaml` by `via`

**Comment by OlivierNicole** [comment]
I'm working on a bytecode-only install (for compiler experimentation) and I ran through this problem as well.
In `$OPAMSWITCH/lib/findlib.conf`:
```
ocamlc="ocamlc.opt"
ocamlopt="ocamlopt.opt"
ocamldep="ocamldep.opt"
ocamldoc="ocamldoc.opt"
```

**Comment by dbuenzli** [comment]
I confirm this still doesn't work with the latest `ocamlfind 1.6.2`, despite [this line](https://github.com/ocaml/opam-repository/blob/master/packages/ocamlfind/ocamlfind.1.6.2/opam#L11) having been added. 

This can be easily checked as follows: 

```
opam switch  4.02.3+bytecode-only
opam install ocamlfind
cat $(opam config var lib)/findlib.conf
destdir="/Users/dbuenzli/.opam/4.02.3+bytecode-only/lib"
path="/Users/dbuenzli/.opam/4.02.3+bytecode-only/lib"
ocamlc="ocamlc.opt"
ocamlopt="ocamlopt.opt"
ocamldep="ocamldep.opt"
ocamldoc="ocamldoc.opt"
```
The result of this is that `ocamlfind ocamlc` picks up the system compiler. 

Not sure whether this is not an upstream bug though.

**Comment by dbuenzli** [comment]
So I tried to compile `ocamlfind` directly from the `tarball` under the `4.02.3+bytecode-only` switch environment and indeed the generated `findlib.conf` has the binaries with `.opt`. 

I also tried to insulate the PATH from the system compiler with the same result (this is a problem we may run quite often with bytecode only switches the native system compiler will be in the PATH, only ocamlc will be overriden by `$(opam config env)`). 

I'll further look into the details of how `findlib.conf` is generated later.

**Comment by dbuenzli** [comment]
It seems I didn't shield the system install correctly. In fact the `findlib.conf` generation is [quite trivial](https://github.com/whitequark/ocaml-findlib/blob/e63c1f1968c29fb573c8651ce62dc931e9c915cc/Makefile#L54-L70), it tries to run the `.opt` versions of the tools and if available mentions them in `findlib.conf`, of course given how PATH is setup we pickup the system ones if the switch ones do not exist.

There are two different problems: 

1. Does ocamlfind actually support the idea of not having a native compiler ? If this is the case, there's unfortunately no variable at the moment in `ocamlc -config` (see [MPR7172](http://caml.inria.fr/mantis/view.php?id=7172)) and we are left with discovery hacks: checking whether ` $(ocamlc -where)/libasmrun.$(EXT_LIB)` exists answers the question and should be communicated to `ocamlfind`.
2. How to discover whether the toolchain was built to native code or not (i.e. `.opt` versions of the toolchain are available). It seems we can't get this information at all, everybody (`opam` included for determining the `ocaml-native-tools` variable) checks whether it finds `.opt` versions of the tools in the PATH, which is of course is completely broken as it will eventually pickup a system compiler if it exists. Assuming bytecode is always installed and cannot be disabled, what should be done is do check whether `$(dirname $(command -v ocamlc))/ocaml{c,opt,dep,doc}.opt` exist.

**Comment by dbuenzli** [comment]
I opened an issue https://gitlab.camlcity.org/gerd/lib-findlib/issues/5

**Comment by dbuenzli** [comment]
In fact 2. should be sufficient it seems that not mentioning the tool in `findlib.conf` results in e.g (here a `findlib.conf` file that has no `ocamldoc` line):
```
$ ocamlfind ocamldoc
ocamlfind: Not supported in your configuration: ocamldoc
```

**Comment by dbuenzli** [comment]
Or maybe not, according to [docs](http://projects.camlcity.org/projects/dl/findlib-1.6.2/doc/ref-html/r825.html) mentioning binaries is optional in `findlib.conf`.

**Comment by OlivierNicole** [comment]
Editing `findlib.conf` by hand doesn't seem to suffice. I have an `ocamlc` and an `ocamlopt` but no `ocamlc.opt`, `ocamlopt.opt` etc. and opam tries to compile `ocamlbuild` using `ocamlc.opt`, because for some reason `OCAML_NATIVE_TOOLS` in ocamlbuild's `Makefile.config` is set to true.

**Comment by dbuenzli** [comment]
A fix seems to be on its way https://gitlab.camlcity.org/gerd/lib-findlib/issues/5

**Comment by dbuenzli** [comment]
The fix made it into ocamlfind 1.7.0 which solves at least point 2. [here](https://github.com/ocaml/opam-repository/issues/1680#issuecomment-221130063)

**Comment by dbuenzli** [comment]
It seems that the latest version of the package [solves](https://github.com/ocaml/opam-repository/blob/master/packages/ocamlfind/ocamlfind.1.7.3/opam#L11) the initial issue via the `ocaml-native` variable. So this I think this can be closed.
